Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -2202,12 +2202,111 @@ def visit_FuncDefNode(self, node):
+        exclude_at_first = ['bases', 'metaclass', 'target']
+        attrs = [ attr for attr in node.child_attrs if attr not in exclude_at_first ]
+
+        prev = self.last_class_name
+        self.last_class_name = name
+        self.visitchildren(node, attrs=attrs)
+        self.last_class_name = prev
+
+        self.visitchildren(node, attrs=exclude_at_first)
+        return node
+
+    def visit_ModuleNode(self, node):
+        dont_mangle_pyx = self.dont_mangle_pyx
+        # classes defined in utility code are treated slightly and attributes
+        # starting with __pyx_ aren't mangled (otherwise memoryviews break)
+        self.dont_mangle_pyx = isinstance(node.scope, NonManglingModuleScope)
+        self.visitchildren(node)
+        self.dont_mangle_pyx = dont_mangle_pyx
+        return node
+
+    def visit_PyClassDefNode(self, node):
+        return self._visit_class_common(node, node.name)
+
+    def visit_CClassDefNode(self, node):
+        # although neither metaclass or target exist for CClassDefNode this
+        # does not prevent them being treated in the same way
+        if node.visibility == "extern": # "public" too?
+            # it's being imported from elsewhere so names are already set
+            return node
+        else:
+            return self._visit_class_common(node, node.class_name)
+
+    def visit_NameNode(self, node):
+        node.name = self.mangle_special_name(node.name)
+        self.visitchildren(node)
+        return node
+
+    def visit_CNameDeclaratorNode(self, node):
+        return self.visit_NameNode(node)
+
+    def visit_DefNode(self, node):
+        return self.visit_NameNode(node)
+
+    def visit_AttributeNode(self, node):
+        node.attribute = self.mangle_special_name(node.attribute)
+        self.visitchildren(node)
+        return node
+
+    def visit_CompilerDirectivesNode(self, node):
+        # It probably isn't desirable to wait until ForwardDeclareTypes to
+        # mangle the names (the earlier the better as far as possible).
+        # Therefore, attempt to change the relevant compiler directives
+        directives = node.directives
+        locals_ = directives.get('locals', None)
+        if locals_:
+            new_locals = { self.mangle_special_name(k): v for k, v in locals_.items() }
+            directives['locals'] = new_locals
+        self.visitchildren(node)
+        return node
+
+    def visit_CallNode(self, node):
+        # Attempt to catch "declare"
+        name = node.function.as_cython_attribute()
+        if name == "declare":
+            _, kwds = node.explicit_args_kwds()
+            if kwds:
+                for var, _ in kwds.key_value_pairs:
+                    var.value = self.mangle_special_name(var.value)
+        self.visitchildren(node)
+        return node
+
+    def mangle_special_name(self, name):
+        pyx_test = ((not self.dont_mangle_pyx) and
+                    (not name.lower().startswith(Naming.pyrex_prefix)))
```

Comment:
I find this kind of double negation ("not don't and not") really hard to read. Is it the same as this?
`is_internal_name = self.dont_mangle_pyx and name.lower().startswith(Naming.pyrex_prefix)`
and then
`if (self.last_class_name and not is_internal_name and …):`
(And I would rather move the "internal" condition behind the `name.startswith("__")` test, which seems the most likely to fail.)